engine-runner and language-server are separate JPMS modules

Akirathan opened the pull request:
Closes #10157

### Pull Request Description

Transform all the transitive dependencies of `language-server` and `runtime` into JPMS modules.

### Important Notes

- We no longer have fat `runtime.jar` and `runner.jar`. Instead, we produce a lot of modularized, small, jar archives into the `component` directory.
  - `runtime-fat-jar` sbt project is removed.
- A lot of mixed Scala/Java sbt projects exclude `module-info.java` from compilation. Therefore, `sbt` cannot track the modifications of those files. If you modify that file, make sure to clean the appropriate project.
- All the sbt projects transitively required by `runtime` and `language-server` are now JPMS modules - their classes contain `module-info.class`.
  - Some problematic dependencies, like `zio` are wrapped in special `zio-wrapper` projects that assemble all the classes from zio and provide our own `module-info.java`. This is, by the way, the recommended approach for modularization.
- Tests (e.g. `runtime-integration-tests` or `language-server/test`) follow the standard approach when testing with JPMS modules - test classes are patched into the corresponding module, and the corresponding module is set to read `ALL-UNNAMED`.
  - This is what Maven surefire plugin does out of the box, when it detects JPMS modules.
- See [JPMSPlugin](https://github.com/enso-org/enso/blob/04495ec03e2b6278029fde9a1e7e9d1306abf3e4/project/JPMSPlugin.scala) for documentation about how the JPMS-related stuff is handled by sbt. 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [X] The documentation has been updated, if necessary.
- [X] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [X] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [X] Unit tests have been written where possible.

Commit:
Merge remote-tracking branch 'refs/remotes/origin/wip/akirathan/improve-jpms-plugin' into wip/akirathan/10157-lang-server-module

Commit:
Use moduleDependencies instead of modulePath

Commit:
Fix compilation of editions

Commit:
Fix compilation of distribution-manager

Commit:
polyglot-api needs to explicitly compile module-info

Commit:
Fix compilationOrder in library-manager and edition-updater

Commit:
engine-runner-common is module

Commit:
JPMSPlugin provides default implementation of compileModuleInfo

Commit:
Remove unused setting key from JPMSUtils.compileModuleInfo

Commit:
JPMSPlugin has internalModuleDependencies and exportedModule tasks.

Commit:
Use BuildVersion instead of buildInfo

Commit:
Manual compilation of module-info.java is reported as warning

Commit:
Define org.enso.scalalibs.wrapper meta project.

Commit:
Fix module check in JPMSPlugin.

This is a fix for projects that declare `Compile /exportJars := true`

Commit:
version-output is a module

Commit:
ydoc-server uses internalModuleDependencies

Commit:
persistance is module

Commit:
engine-common uses internalModuleDependencies

Commit:
polyglot-api does not override compileModuleInfo task

Commit:
runtime-parser uses internalModuleDependencies

Commit:
edition-updater is module

Commit:
Update moduleDependencies for distribution-manager

Commit:
editions is module

Commit:
Fix some dependencies of modules

JaroslavTulach reviewed (approved):
Overall the change looks good.

Commit:
scala-yaml is a module

Commit:
Add scala-compiler to scala-libs-wrapper

Commit:
cli depends on scala-library module

Commit:
Add dependencies for distribution-manager module

Commit:
Add some scala-library dependencies in some modules

Commit:
engine-runner uses internalModuleDependencies

Commit:
Fix module dependencies of library-manager

Commit:
Rename org.enso.scalalibs.wrapper to org.enso.scala.wrapper

Commit:
Add jsoniter-scala-macros to org.enso.scala.wrapper fat module

Commit:
Fix dependencies of some projects

Commit:
polyglot-api does not depend on truffle-api

Commit:
Fix dependencies of some projects

Commit:
runtime does not use com.google.common

Commit:
runtime is a module

Commit:
text-buffer is a module

Commit:
refactoring-utils is a module

Commit:
runtime-compiler is a module

Commit:
runtime-instrument-common is a module

Commit:
connected-lock-manager is a module

Commit:
JPMSUtils reports project name in some error messages

Commit:
Modularize some instruments

Commit:
module-info compilation is cached

Commit:
runtime-instrument-runtime-server is module

Commit:
runtime-language-epb is module

Commit:
Remove runtime-fat-jar

Commit:
engine-runner is not a fat jar

Commit:
JPMSPlugin defines exportedModuleBin task

Commit:
Redefine componentModulesPaths task

Commit:
interpreter-dsl is module

Commit:
Redefine componentModulesPaths task

Commit:
fmt sbt

Commit:
scala-libs-wrapper is a modular fat jar

Commit:
Add some module deps to org.enso.runtime

Commit:
engine-runner is not a fat jar

Commit:
Rename package in logging-config

Commit:
Rename package in logging-service

Commit:
Rename package in logging-service-logback

JaroslavTulach commented on engine/runner-common/src/main/java/module-info.java:
```diff
@@ -0,0 +1,12 @@
+  requires org.enso.librarymanager;
+  requires org.enso.pkg;
+  requires org.slf4j;
+  requires scala.library;
```
Áááágr:
```
engine/runner-common/src/main/scala/org/enso/runner/common/CompilerBasedDependencyExtractor.scala
engine/runner-common/src/main/scala/org/enso/runner/common/ProfilingConfig.scala
```
Just two classes away from getting rid of another Scala dependency - alas, not: the scal would get in the list of dependencies transitively anyway.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/module-info.java:
```diff
@@ -0,0 +1,17 @@
+  requires org.enso.pkg;
+  requires org.enso.runtime.parser;
+
+  requires org.openide.util.lookup.RELEASE180;
```
Dependencies on `org.openide.util.lookup` are probably just `static` in all cases except `persistance-api`.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/module-info.java:
```diff
@@ -0,0 +1,17 @@
+  exports org.enso.compiler;
+  exports org.enso.compiler.context;
+  exports org.enso.compiler.data;
+  exports org.enso.compiler.dump;
```
Who needs `dump`? I don't want to expose `dump` from the compiler. Definitely not as a stable API as astable as `org.enso.compiler`. Maybe just `export to`.

JaroslavTulach commented on engine/runtime-fat-jar/src/main/java/module-info.java:
No more module info for `runtime-fat-jar`. OK, that's a change!

JaroslavTulach commented on engine/runtime-instrument-runtime-server/src/main/java/module-info.java:
```diff
@@ -0,0 +1,15 @@
+  requires org.enso.distribution;
+  requires org.enso.lockmanager;
+
+  requires org.openide.util.lookup.RELEASE180;
```
Unlikely that `org.openide.util.lookup` would be needed by this instrument.

JaroslavTulach commented on engine/runtime-language-epb/src/main/java/module-info.java:
```diff
@@ -0,0 +1,7 @@
+module org.enso.runtime.language.epb {
+  requires java.logging;
+  requires org.graalvm.truffle;
```
Finally a module that doesn't need Scala!

JaroslavTulach commented on engine/runtime/src/main/java/module-info.java:
```diff
@@ -0,0 +1,43 @@
+  exports org.enso.interpreter.runtime.error;
+  exports org.enso.interpreter.runtime.instrument;
+  exports org.enso.interpreter.runtime.tag;
+  exports org.enso.interpreter.node.expression.builtin.debug;
```
Quite a lot of exports. Who do we export for example these `builtin`s to? I'd be good to tighten the `runtime` API up and export less.

Probably better to explicitly `exports to` just to our instruments. We don't want any 3rd party to use these packgages.

But can wait for later.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/EnsoLanguage.java:
```diff
@@ -27,10 +27,10 @@
+import org.enso.interpreter.runtime.instrument.NotificationHandler;
+import org.enso.interpreter.runtime.instrument.NotificationHandler.Forwarder;
+import org.enso.interpreter.runtime.instrument.NotificationHandler.TextMode$;
+import org.enso.interpreter.runtime.instrument.Timer;
```
Oh yeah. Old sins of mine. Splitting projects without renaming packages fights back in modular world!

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeNode.java:
```diff
@@ -1,6 +1,6 @@
 package org.enso.interpreter.node.expression.builtin.meta;
 
-import com.google.common.base.Objects;
```
Maybe we could get rid of that library all together:

![com.google.common](https://github.com/user-attachments/assets/c0986f67-8133-4e68-aafa-60470cb43998)

JaroslavTulach commented on lib/java/persistance/src/main/java/module-info.java:
```diff
@@ -0,0 +1,6 @@
+module org.enso.persistance {
+  requires org.openide.util.lookup.RELEASE180;
```
This one is probably justified, because we are using `metainfServicesLookup` - problems with that approach are described in #10972

JaroslavTulach commented on lib/java/scala-libs-wrapper/src/main/java/module-info.java:
Cats, cats, cats & Circe and Shapeless... shouldn't shapeless be just a _compile time_ dependency since your last cleanup, @hubertp?

hubertp replied:
No. It's a transitive dependency of other libraries that are being heavily used.

JaroslavTulach commented on lib/scala/cli/src/main/java/org/enso/cli/internal/package-info.java:
You can remove `package-info` if you put some other Java file into the compilation scope. `module-info` is a good replacement.

JaroslavTulach commented on lib/scala/pkg/src/main/java/module-info.java:
```diff
@@ -0,0 +1,12 @@
+  requires org.apache.commons.compress;
+  requires org.enso.editions;
+  requires org.enso.semver;
+  requires org.enso.scala.yaml;
```
Wouldn't it be better to `requires transitive org.enso.scala.yaml` rather then adding the next two libraries here as well?

JaroslavTulach reviewed (approved):
Enormous work and very needed! I hope things don't fall apart and these changes get in soon while keeping Enso working.

Commit:
Fix dependencies of exportedModuleBin task

Commit:
Mixed projects have own compileJava task

this task does not compile only module-info.java but all the java sources. So that we can see errors more easily.

When only module-info.java is compiled, the only errors that we can see are that we did not include some modules on module-path.

Commit:
Fix definition of exportedModule task.

Commit:
Remove usages of non-existing buildInfo and replace it with BuildVersion

Commit:
Fix some dependencies of org.enso.runtime module

Commit:
module-info compilation is handled directly by FrgaalCompiler

Commit:
module-info compilation is forced for projects that has only Scala sources with single module-info.java

Commit:
Fix compilation of org.enso.runtime

Commit:
manual module-info compilation is not a warning

Commit:
Rename packages in logging-utils-akka

Commit:
Create org.enso.language.server.deps.wrapper module

Commit:
language-server is module

Commit:
Creat akka-wrapper modular fat jar

Commit:
fmt

Commit:
Define common settings for modularFatJarWrapper

Commit:
Fix compilation of json-rpc-server

Commit:
Use akka and zio wrappers

Commit:
language-server depends on org.eclipse.jgit

Commit:
Fix some dependencies - update library manifests works now!

Commit:
update library manifests invokes runner directly

Commit:
buildEngineDistribution does not copy runner.jar

Commit:
Remove EngineRunnerBootLoader

Commit:
Fix compilation of std libs

Commit:
--patch-module and --add-exports are also passed to javac

Commit:
Rename package in runtime-integration-tests.

The package name org.enso.compiler clashes with the package from the module

Commit:
Remove usage of buildInfo

Commit:
FrgaalJavaCompiler can deal with non-existing module-info.java when shouldCompileModuleInfo is true.

It just generates a warning in such case as it suggests that there is something wrong with the project configuration.

Commit:
Revert AliasAnalysisTest.scala

Commit:
Fix dependencies and java cmdline options for runtime-integration-tests

Commit:
Rename test package

Commit:
runtime-integration-test depends on logging-service-logback/Test/compile

Commit:
Rename package in logging-service-logback/Test

Commit:
Fix FrgaalJavaCompiler creation for projects

Commit:
Sanitize Test/javaOptions arguments

Commit:
Sanitize Test/javaOptions arguments

Commit:
All the JPMSPlugin settings are scoped

Commit:
Remove unused sbt tasks

Commit:
modularFatJarWrapperSettings do not override javacOptions

Commit:
Resolve issue "Cannot find TestLoggerProvider" in runtime-integration-tests

Commit:
org.enso.runtime module is open

Commit:
Test that test classes are unconditionally opened and exported

Commit:
polyglot-api-macros is a module

Commit:
JPMSPlugin handles --add-opens cmdline option

Commit:
RuntimeServerTest ensures instruments are initialized

Commit:
Add some exports to org.enso.runtime.compiler

Commit:
Add instruments on module-path to runtime-integration-tests

Commit:
Replace TestLogProviderOnClassPath with TestLogProviderOnModulePath

Commit:
Replace buildInfo with BuildVersion

Commit:
Add jpms-wrapper-scalatest

Commit:
ReportLogsOnFailure is in non-modular testkit project

Commit:
Add necessary dependencies to testkit project

Commit:
Revert "Add jpms-wrapper-scalatest"

This reverts commit 732b3427a2c4a7051c831f3439b4fc04ba126ae1.

Commit:
modularize filewatcher and wrap its dependencies

Commit:
Initial fix for language-server/test

Commit:
frgaal compiler setting are scoped for Compile and Test

Commit:
Rename package in language-server/test

Commit:
Exclude com.sun.jna from wrapper jars

Commit:
Rename package in library-manager-test

Commit:
testkit is an automatic module

Commit:
process-utils is module

Commit:
akka-wrapper contains akka-http

Commit:
Some fixes for library-manager-test

Commit:
Fix dependencies for akka-wrapper

Commit:
scala-libs-wrapper exports shapeless

Commit:
lang server deps wrapper exports pureconfig

Commit:
json-rpc-server requires org.slf4j

Commit:
Add some dependencies

Commit:
lang server deps wrapper exports pureconfig.generic

Commit:
language server test requires bouncycastle provider

Commit:
language server depends on cli

Commit:
directory-watcher wrapper requires org.slf4j

Commit:
WatcherAdapter logs unsuccessful initialization errors

Commit:
Fix error reporting in WatcherAdapter

Commit:
Fix rest of the language-server tests

Commit:
language-server-deps-wrapper depends on scala-libs-wrapper

Commit:
Fix rest of the language-server tests

Commit:
Missing module-info.class in an internal project is a warning, not an error

Commit:
Rename jpms-methvin-directory-watcher-wrapper to a simpler name

Commit:
compileOrder has to be specified before libraryDependencies

Commit:
exclude module-info.java from polyglot-api-macros

Commit:
Remove temporary logging in customFrgaalCompilerSettings

Commit:
Fix compilation of logging-service-logback

Commit:
Fix compilation of runtime-benchmarks

Commit:
Fix runtime-benchmarks/run

Commit:
HostClassLoader delegates to org.graalvm.polyglot class loader if org.enso.runtime is not on boot layer

Commit:
org.enso.runtime.lnaguage.epb module must be opened to allow it to be used by annnotation processor

Commit:
fmt

Commit:
Merge branch 'develop' into wip/akirathan/10157-lang-server-module

# Conflicts:
#	build.sbt
#	engine/language-server/src/main/scala/org/enso/languageserver/boot/MainModule.scala
#	engine/runtime-instrument-repl-debugger/src/main/java/org/enso/interpreter/instrument/repl/debugger/ReplDebuggerInstrument.java
#	project/JPMSPlugin.scala
#	project/JPMSUtils.scala

JaroslavTulach commented:
[runtime-fat-jar not found](https://github.com/enso-org/enso/actions/runs/10696346989/job/29651364077?pr=10823#step:5:20)

Commit:
Fix afetr merge

Commit:
Add module deps after merge

Commit:
Print stack trace of the uncaught exception from the annotation processor

Commit:
Merge branch 'develop' into wip/akirathan/10157-lang-server-module

# Conflicts:
#	build.sbt
#	docs/infrastructure/logging.md
#	engine/runtime-fat-jar/src/main/java/module-info.java
#	engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/TestLogProviderOnClasspath.java
#	lib/scala/version-output/src/main/java/org/enso/version/VersionDescription.java

Commit:
Remove akka-actor-typed from akka-wrapper

Commit:
runtime-instrument-common depends on slf4j

Commit:
Fix module-path for runtime-instrument-repl-debugger

JaroslavTulach commented on lib/java/akka-wrapper/src/main/java/module-info.java:
```diff
@@ -0,0 +1,156 @@
+  exports akka.util.ccompat;
+
+  // akka-actor-typed
+  exports akka.actor.typed;
```
Hasn't `akka...typed` been removed in #10905?

Akirathan replied:
Yes, `exports akka.actor.typed.*` has been already removed in the latest revision on this PR.

JaroslavTulach commented on distribution/bin/enso:
```diff
@@ -1,5 +1,5 @@
 
 JAVA_OPTS="--add-opens=java.base/java.nio=ALL-UNNAMED $JAVA_OPTS"
-exec java --module-path $COMP_PATH $JAVA_OPTS -m org.enso.runtime/org.enso.EngineRunnerBootLoader "$@"
+exec java --module-path $COMP_PATH $JAVA_OPTS -m org.enso.runner/org.enso.runner.Main "$@"
```
Now the runner is on `--module-path` without a single line of change in `org.enso.runner.Main` - interesting. We'll see what happens with this when #10010 gets closer.

JaroslavTulach commented on engine/language-server/src/main/java/module-info.java:
```diff
@@ -0,0 +1,41 @@
+  requires org.apache.commons.io;
+  requires org.graalvm.polyglot;
+  requires org.eclipse.jgit;
+  requires org.openide.util.lookup.RELEASE180;
```
Very likely unnecessary

JaroslavTulach reviewed (approved):
Please remove runtime dependencies on `org.openide.util.lookup`.

Commit:
runtime-benchmarks depends on runtime-language-arrow

Commit:
--module-path is passed directly to frgaal

Commit:
Fix some module-related cmd line options for std-benchmarks

Commit:
Revert "--module-path is passed directly to frgaal"

This reverts commit da63f66a0ed2377f78d90d13bdd955f903f0ff19.

JaroslavTulach committed:
Avoid closing of System.err when closing Context

JaroslavTulach committed:
Avoid processing altogether when requested annotations are empty

Commit:
Pass shouldNotLimitModules opt to frgaal

Commit:
Pass module-path and add-modules options with -J prefix to frgaal

Commit:
BenchProcessor annotation processor creates its own truffle module layer

Commit:
bench-processor and benchmarks-common are modules

Commit:
fmt

Commit:
Merge branch 'develop' into wip/akirathan/10157-lang-server-module

# Conflicts:
#	lib/scala/bench-processor/src/main/java/org/enso/benchmarks/processor/BenchProcessor.java

Commit:
Fix after mege

Commit:
Enable JMH annotation processor

Commit:
Fix compileOrder in some projects

Commit:
Insert TruffleBoundary to QualifiedName.

This is a revert

Commit:
Fix building of engine-runner native image

Commit:
Add more deps to the native image

Commit:
Force module-info compilation in instruments.

This fixes some weird sbt bug

Commit:
Don't run engine-runner/assembly from Rust build script

Commit:
Merge branch 'develop' into wip/akirathan/10157-lang-server-module

Commit:
Update docs of JPMSPlugin

Commit:
fmt

Commit:
runtime-benchmarks depends on benchmarks-common module

Commit:
Fix benchmark report writing

Commit:
std-benchmarks annot processing does not take settings from runtime-benchmarks

Commit:
Suppress interpreter only warning in annotation processor

Commit:
Runtime version manager does not expect runtime.jar fat jar

Commit:
fmt

JaroslavTulach commented:
[Failures in JVM tests](https://github.com/enso-org/enso/actions/runs/10920765574/job/30311358097?pr=10823#step:7:3467) like:
```
The runner used to create the project returned exit code 1.
Runner's output: stdout> Error occurred during initialization of boot layer
stdout> java.lang.module.FindException: Error reading 
   /var/folders/3p/1bg9mx4130dgqr_85c65ptwh0000gn/T/13765391451016182419/test_data/dist/0.0.1/component/runtime.jar\nstdout> Caused by: java.util.zip.ZipException: zip END header not found
```
Benchmarks [do not compile](https://github.com/enso-org/enso/actions/runs/10920765574/job/30311357782?pr=10823#step:7:2105):
```
Module org.apache.commons.io reads more than one module named org.graalvm.nativeimage
```

Commit:
Fix module entry point

Commit:
Move some polyglot tests to runtime-integration-tests.

Also make their output silent

Commit:
pkg has no dependency on org.graalvm.truffle

Commit:
Fix compiler dependencies test

Commit:
Rename all runtime.jar in fake releases

Commit:
Add language-server with dependencies to component dir

Commit:
No module-info.class in target dir is warning not error

Commit:
language-server does not depend on netbeans lookup uitl

Commit:
Declare LanguageServerApi service provider in module-info

Commit:
connected-lock-manager-server is JPMS module

Commit:
task-progress-notifications is module

Commit:
Add fansi-wrapper module

Commit:
Fix compilation of connected-lock-manager-server

Commit:
Define correct Test/internalModuleDependencies for project-manager

Commit:
fmt

Commit:
Fix LauncherRunnerSpec - no runtime.jar

Commit:
Add fansi-wrapper to runtime-integration-tests and runtime-benchmarks

Commit:
Fix engine-runner native image build

Commit:
Use newer JNA version - fixes running of hyperd

hubertp committed:
DRY

hubertp committed:
scala-compiler DRY

hubertp committed:
fmt

hubertp committed:
More build.sbt refactoring

Commit:
Include runtime-instrument-id-execution in engine-runner native image

Commit:
TruffleBoundary for QualifiedName.toString

hubertp committed:
Finding a needle in a haystack

:facepalm:

hubertp commented on lib/java/akka-wrapper/src/main/java/module-info.java:
```diff
@@ -0,0 +1,139 @@
+  // io.spray:spray-json_2.13
+  exports spray.json;
+
+  // scala-lang.modules:scala-arser-combinators_2.13:1.1.2
```
```suggestion
  // scala-lang.modules:scala-parser-combinators_2.13:1.1.2
```

hubertp commented on lib/java/akka-wrapper/src/main/java/module-info.java:
```diff
@@ -0,0 +1,139 @@
+  exports spray.json;
+
+  // scala-lang.modules:scala-arser-combinators_2.13:1.1.2
+  // Loaded by classes from com.typesafe.sslconfig.ssl.*
```
:(

hubertp committed:
More scala-library DRY

hubertp committed:
more mixed-java/scala goodies

Commit:
Merge branch 'develop' into wip/akirathan/10157-lang-server-module

Commit:
Fix compilation of syntax-rust-definition

Commit:
Test that engine-runner does not depend on language-server

hubertp committed:
Append rather than assign `moduleDependencies`

`++=` is less error prone than `:=`. Also discovered some unnecessary
dependencies.

Commit:
Replace : with File.pathSeparator

JaroslavTulach commented on engine/runner/src/test/java/org/enso/runner/EngineRunnerDependenciesTest.java:
```diff
@@ -16,9 +16,6 @@ public class EngineRunnerDependenciesTest {
-  // ignored for now as engine-runner still has "Runtime" dependency on language-server and that one
-  // appears in test classpath - remove once language-server is packaged as its own JAR file
-  @Ignore
   public void unableToLoadClassFromLanguageServerProject() {
```
+1

Commit:
[WIP] Make logging in ProjectService more verbose

JaroslavTulach committed:
language-server/test didn't start because of missing lookup and fansi modules

JaroslavTulach committed:
Formatting

JaroslavTulach committed:
org.enso.cli.task.notifications needs Akka and Circe to link

hubertp committed:
project-manager/test depends on buildEngineDistribution

Commit:
[WIP] Even more verbose logging for creating projects

Commit:
Merge remote-tracking branch 'origin/wip/akirathan/10157-lang-server-module' into wip/akirathan/10157-lang-server-module

Commit:
[WIP] Even more verbose logging for creating projects

Commit:
Revert "[WIP] Even more verbose logging for creating projects"

This reverts commit a7067c8472c34b6934a0ea6f4d8075628eaf14e6.

Commit:
Revert "[WIP] Even more verbose logging for creating projects"

This reverts commit fc6f53d4f1dd4a86a7b01e9c4f86b35a824bf88b.

Commit:
Revert "[WIP] Make logging in ProjectService more verbose"

This reverts commit 427428e1420984c5faee0ce0981a3b405964df47.

Commit:
All the project with JPMSPlugin has stripped artifact names

Commit:
Revert all placeholder fake release components to runtime.jar without version

hubertp committed:
Eliminate a cross version hack

We shouldn't be specifying Scala dependencies with a Scala cross version
in the suffix.

hubertp committed:
Address SBT lint warnings

Commit:
Revert "Eliminate a cross version hack"

This reverts commit 8861dab288cbb0ff7f2f456c8725d75d4442aa9c.

Commit:
logging-service-logback  is mixedJavaScalaProject

Commit:
fmt

hubertp commented on distribution/bin/enso:
```diff
@@ -1,5 +1,5 @@
 COMP_PATH=$(dirname "$0")/../component
```
This file could be auto-generated so that we don't hardcode here the name of the module. Not for this PR, though.

hubertp commented on distribution/bin/enso.bat:
```diff
@@ -1,5 +1,5 @@
 set comp-dir=%~dp0\..\component
 set JAVA_OPTS=%JAVA_OPTS% --add-opens=java.base/java.nio=ALL-UNNAMED
-java --module-path %comp-dir% -Dpolyglot.compiler.IterativePartialEscape=true %JAVA_OPTS% -m org.enso.runtime/org.enso.EngineRunnerBootLoader %*
+java --module-path %comp-dir% -Dpolyglot.compiler.IterativePartialEscape=true %JAVA_OPTS% -m org.enso.runner/org.enso.runner.Main %*
```
Same comment about auto-generation

hubertp commented on engine/runtime-instrument-common/src/main/java/module-info.java:
```diff
@@ -0,0 +1,29 @@
+
+  requires org.graalvm.truffle;
+  requires org.graalvm.polyglot;
+  // TODO: com.oracle.truffle.api.TruffleContext?
```
spurious?

hubertp commented on engine/runtime-integration-tests/src/test/scala/org/enso/polyglot/test/ApiTest.scala:
```diff
@@ -1,14 +1,17 @@
 
 class ApiTest extends AnyFlatSpec with Matchers {
   import org.enso.common.LanguageInfo._
+  val out = new ByteArrayOutputStream()
```
Do we do anything with it? We certainly don't close it.

Akirathan replied:
The whole `executionContext` is not closed in these tests. What's the point of closing `ByteArrayOutputStream`? It will get garbage collected sooner or later. And we don't do anything with it on purpose, this is equivalent to setting `/dev/null` as the output.

hubertp commented on engine/runtime-integration-tests/src/test/scala/org/enso/polyglot/test/ModuleManagementTest.scala:
```diff
@@ -28,6 +29,7 @@ class ModuleManagementTest
         packageName,
         "Enso_Test"
       )
+    val out = new ByteArrayOutputStream()
```
Will probably lead to leaks unless you close it.

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/HostClassLoader.java:
```diff
@@ -50,7 +58,7 @@ protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundE
       return l;
     }
-    if (ClassLoaderConstants.CLASS_DELEGATION_PATTERNS.stream().anyMatch(name::startsWith)) {
+    if (!isRuntimeModInBootLayer && name.startsWith("org.graalvm")) {
```
Could we avoid hardcoding that? Say by `.getClass` and retrieving the package

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/HostClassLoader.java:
```diff
@@ -28,8 +25,19 @@ final class HostClassLoader extends URLClassLoader implements AutoCloseable {
+  static {
+    var bootModules = ModuleLayer.boot().modules();
+    var hasRuntimeMod =
+        bootModules.stream().anyMatch(module -> module.getName().equals("org.enso.runtime"));
```
Can we avoid this hardcoded package name somehow?

hubertp commented on lib/scala/interpreter-dsl/src/main/java/module-info.java:
```diff
@@ -0,0 +1,9 @@
+  requires com.google.common;
+  requires java.compiler;
+  requires org.apache.commons.lang3;
+  requires org.openide.util.lookup.RELEASE180;
```
`requires static`?

hubertp commented on lib/scala/project-manager/src/test/java/org/enso/projectmanager/TestLogProviderOnClasspath.java:
```diff
@@ -12,8 +12,9 @@
- * In the `runtime/Test` testing suite, {@link }org.enso.logger.TestLogProvider} should be among the
- * logging providers, because it is explicitly chosen as the logging provider for the tests.
+ * In the `runtime/Test` testing suite, {@link
+ * }org.enso.logging.service.logback.test.provider.TestLogProvider} should be among the logging
```
```suggestion
 * org.enso.logging.service.logback.test.provider.TestLogProvider} should be among the logging
```
typo, right?

Commit:
Stripped artifact name contains classifier.

This fixes tests as those were named like `artifact-tests.jar`.

hubertp committed:
Don't use LocalProject unless really needed

JaroslavTulach commented:
Strange [failure on Windows](https://github.com/enso-org/enso/actions/runs/11017283297/job/30594798699?pr=10823#step:7:2082) while Linux is OK. Problem with `\`?
```
  INFO ide_ci::program::command: sbt ℹ️ [warn] An annotation processor threw an uncaught exception.
 INFO ide_ci::program::command: sbt ℹ️ [warn] Consult the following stack trace for details.
 INFO ide_ci::program::command: sbt ℹ️ [warn] java.lang.module.ResolutionException: Module org.apache.commons.io reads more than one module named org.graalvm.collections
 INFO ide_ci::program::command: sbt ℹ️ [warn] 	at java.base/java.lang.module.Resolver.resolveFail(Resolver.java:900)
 INFO ide_ci::program::command: sbt ℹ️ [warn] 	at java.base/java.lang.module.Resolver.checkExportSuppliers(Resolver.java:721)
 INFO ide_ci::program::command: sbt ℹ️ [warn] 	at java.base/java.lang.module.Resolver.finish(Resolver.java:380)
 INFO ide_ci::program::command: sbt ℹ️ [warn] 	at java.base/java.lang.module.Configuration.<init>(Configuration.java:139)
 INFO ide_ci::program::command: sbt ℹ️ [warn] 	at java.base/java.lang.module.Configuration.resolve(Configuration.java:422)
 INFO ide_ci::program::command: sbt ℹ️ [warn] 	at java.base/java.lang.module.Configuration.resolve(Configuration.java:254)
 INFO ide_ci::program::command: sbt ℹ️ [warn] 	at org.enso.benchmarks.processor.BenchProcessor.createLayerFromPaths(BenchProcessor.java:141)
 INFO ide_ci::program::command: sbt ℹ️ [warn] 	at org.enso.benchmarks.processor.BenchProcessor.createTruffleModuleLayer(BenchProcessor.java:110)
 INFO ide_ci::program::command: sbt ℹ️ [warn] 	at org.enso.benchmarks.processor.BenchProcessor.process(BenchProcessor.java:192)
```

JaroslavTulach commented:
One `runtime-integration-test` failure out of [three thousands](https://github.com/enso-org/enso/actions/runs/11017283297/job/30594799454?pr=10823#step:7:6071)!
```
 INFO ide_ci::program::command: sbt ℹ️ [error] Failed: Total 3006, Failed 1, Errors 0, Passed 3005, Ignored 23, Pending 30
 INFO ide_ci::program::command: sbt ℹ️ [error] Failed tests:
 INFO ide_ci::program::command: sbt ℹ️ [error] 	org.enso.interpreter.test.TestLogProviderOnModulePath

  INFO ide_ci::program::command: sbt ℹ️ [error] Test org.enso.interpreter.test.TestLogProviderOnModulePath.testLogProviderOnClasspath failed: java.lang.AssertionError: 
 INFO ide_ci::program::command: sbt ℹ️ [error] Expected: is <true>
 INFO ide_ci::program::command: sbt ℹ️ [error]      but: was <false>, took 0.003 sec
 INFO ide_ci::program::command: sbt ℹ️ [error]     at org.enso.interpreter.test.TestLogProviderOnModulePath.testLogProviderOnClasspath(TestLogProviderOnModulePath.java:18)
 INFO ide_ci::program::command: sbt ℹ️ [error]     at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
 INFO ide_ci::program::command: sbt ℹ️ [error]     at java.lang.reflect.Method.invoke(Method.java:580)
```

JaroslavTulach commented:
Licenses review is finally a bit useful:
```
There are 15 fatal-errors found in the review.
License review file tools/legal-review/launcher/reviewed-licenses/The_BSD_License is missing. Either review the default license or set a `custom-license` for packages that used it.
15 copyrights are not reviewed in net.java.dev.jna # jna # 5.14.0.
No files or copyright information are included for net.java.dev.jna # jna # 5.14.0. Generally every dependency should have _some_ copyright info, so this suggests all our heuristics failed. Please find the information manually and add it using `files-add` or `copyright-add`. Even if the dependency is in public domain, it may be good to include some information about its source.
1 files are not reviewed in com.github.plokhotnyuk.jsoniter-scala # jsoniter-scala-macros_2.13 # 2.28.5.
No files or copyright information are included for com.github.plokhotnyuk.jsoniter-scala # jsoniter-scala-macros_2.13 # 2.28.5. Generally every dependency should have _some_ copyright info, so this suggests all our heuristics failed. Please find the information manually and add it using `files-add` or `copyright-add`. Even if the dependency is in public domain, it may be good to include some information about its source.
1 files are not reviewed in com.github.plokhotnyuk.jsoniter-scala # jsoniter-scala-core_2.13 # 2.28.5.
No files or copyright information are included for com.github.plokhotnyuk.jsoniter-scala # jsoniter-scala-core_2.13 # 2.28.5. Generally every dependency should have _some_ copyright info, so this suggests all our heuristics failed. Please find the information manually and add it using `files-add` or `copyright-add`. Even if the dependency is in public domain, it may be good to include some information about its source.
7 copyrights are not reviewed in io.github.java-diff-utils # java-diff-utils # 4.12.
No files or copyright information are included for io.github.java-diff-utils # java-diff-utils # 4.12. Generally every dependency should have _some_ copyright info, so this suggests all our heuristics failed. Please find the information manually and add it using `files-add` or `copyright-add`. Even if the dependency is in public domain, it may be good to include some information about its source.
2 files are not reviewed in org.scala-lang # scala-compiler # 2.13.11.
12 copyrights are not reviewed in org.scala-lang # scala-compiler # 2.13.11.
No files or copyright information are included for org.scala-lang # scala-compiler # 2.13.11. Generally every dependency should have _some_ copyright info, so this suggests all our heuristics failed. Please find the information manually and add it using `files-add` or `copyright-add`. Even if the dependency is in public domain, it may be good to include some information about its source.
19 copyrights are not reviewed in org.jline # jline # 3.23.0.
No files or copyright information are included for org.jline # jline # 3.23.0. Generally every dependency should have _some_ copyright info, so this suggests all our heuristics failed. Please find the information manually and add it using `files-add` or `copyright-add`. Even if the dependency is in public domain, it may be good to include some information about its source.
Default license The BSD License for org.jline # jline # 3.23.0 is used, but that license is not reviewed (need to add an entry to `reviewed-licenses`).
```
I don't think we want to distribute `scala-compiler`, neither `net.java.dev.jna`, etc.

Commit:
Add more logging when BenchProcessor fails

Commit:
logging-service-logback is not mixed project

Akirathan commented:
Looking at our old `runtime.jar` and `runner.jar` fat jars and all the dependencies now reported by the license verification tool:
- `org.scala-lang:scala-reflect` **is included in `runtime.jar`** (for example package `scala.reflect.runtime`)
- `net.java.dev.jna:jna` (Automatic module `com.sun.jna`) **is included in `runner.jar`**, as it is a transitive dependency of `io.methvin:directory-watcher`.
- `com.github.plokhotnyuk.jsoniter-scala:jsoniter-scala-core` is **included in runtime.jar**.
- `org.jline:jline` is **included in runner.jar**.
- `org.scala-lang:scala-compiler` (along with its dependency `com.github.difflib`) is not in runtime.jar, neither runner.jar.

Obviously, license verification tool is not perfect and it has not reported, e.g., `net.java.dev.jna` despite the fact that we have included that in our `runner.jar`. Nevertheless, the only one additional dependency that is now distributed in the `component` directory is `org.scala-lang:scala-compiler` which is a dependency of `org.enso.scala.wrapper` module (I think this is because of `io.circe` or other super fun Scala stuff). I will try once more to remove the `org.scala-lang:scala-compiler` dependency from our distribution, but all the other dependencies shall be distributed, as they always were distributed. We just need to fix the appropriate records in the license verification tool.

Note that most probably, the license verification tool uses something like `Compile / fullClasspath` for collecting the dependencies. Which is very wrong. This is not what is distributed at all now. All the JARs that are distributed are collected by the `componentModulesPaths` task (you can evaluate it by `print componentModulesPaths`, or simply by checking what is in the `built-distribution/.../component` directory).

JaroslavTulach commented:
> only one additional dependency ... in the `component` directory is `org.scala-lang:scala-compiler` ... a dependency ... because of `io.circe` 

I believe the compiler is only needed during compile time - e.g. it would be enough to declare it as `"provided"`. I was thinking of splitting the `scala-libs-wrapper` into two parts - runtime and compile time, but they I messed it up and gave up.

JaroslavTulach committed:
Work with java.io.File.getPath to avoid mixing slash and backslash on Windows

hubertp committed:
Reapply "Eliminate a cross version hack"

This reverts commit edaa436ee89670f4065c13dda7593e19fb207b10.

hubertp committed:
Pass scalaBinaryVersion correctly

Commit:
Remove scala-compiler from the distribution

Commit:
Fix IllegalAccessErrors from serde

Commit:
typos

Commit:
License review

Commit:
fmt

Commit:
Move testLogProviderOnModulePath to TestJPMSConfiguration

Commit:
logging-service-logback is not a mixed project

JaroslavTulach commented on distribution/project-manager/THIRD-PARTY/net.java.dev.jna.jna-5.14.0/NOTICES:
```diff
@@ -0,0 +1,29 @@
+/* Copyright (c) 2012 Timothy Wall, All Rights Reserved
```
I am pretty sure we shall not be using JNA in `project-manager` in runtime. `project-manager` is _native image_ compiled application and it is using [system Java](https://www.graalvm.org/sdk/javadoc/org/graalvm/nativeimage/c/package-summary.html) and not JNA.

Akirathan replied:
Seems to be caused by unfortunately specified dependencies of `scala-libs-wrapper`:
```
project-manager --> scala-yaml --> scala-libs-wrapper --> net.java.dev.jna
```

But `net.java.dev.jna` is excluded from `scala-libs-wrapper.jar`. So it is **not** present in project-manager.

radeusgd commented on lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/runner/Runner.scala:
```diff
@@ -197,29 +197,14 @@ class Runner(
       val componentPath = engine.componentDirPath.toAbsolutePath.normalize
-      var jvmArguments =
-        manifestOptions ++ environmentOptions ++ commandLineOptions
-      if (shouldInvokeViaModulePath) {
```
I think ~~removing this conditional was a mistake~~ we should have not removed this conditional from the logic - this causes newer launcher to be unable to run older Engine versions that relied on the `-jar` dispatch.

Similarly, the change
```diff
- org.enso.runtime/org.enso.EngineRunnerBootLoader
+ org.enso.runner/org.enso.runner.Main
```
should also be guarded by a version check - the line to use should depend on the engine version.
Alternatively, this line should have been moved to the `jvm-options` section of the `manifest.template.yaml` to be able to specify it with each engine version without such conditionals.

Currently it seems that launcher released after 25th of September will not be able to run earlier releases. I'm not sure if it is worth adding fixes to newer launcher to support these old releases - I'm not aware of anyone relying on these old versions, so it perhaps seems better to keep the logic simple and focus on avoiding breakage in future releases. Unless someone thinks that we should make the new launcher work with older versions, then please create a relevant ticket.

Nonetheless, for future we should be careful when changing how the engine is started. The `ensoup` launcher has numerous facilities (including self-upgrade mechanism) to ensure no compatibility breakage if releases are coordinated accordingly. Let's use them next time.

Perhaps we may need to think of some way of testing the logic changes - running new engine with old launcher and vice-versa. A problem is that the logic normally works on **released** releases making it hard to test (there are unit tests that test self-upgrade etc. but not this cross-version compatibility). Perhaps next time a change like this is applied, we should think of creating test scenarios that allow us to test it.

cc: @JaroslavTulach @Akirathan

JaroslavTulach replied:
Yeah, neither me nor @Akirathan realized we are facing an instance of [Incremental Deployment](http://wiki.apidesign.org/wiki/Incremental_deployment) problem when it comes to communication between `ensoup` and `enso`. Let's keep it in mind for the next time.

Moreover, as [described here](https://github.com/enso-org/enso/issues/11794#issuecomment-2527491882) - let's move away from using the `org.enso.runner/org.enso.runner.Main` class name for communication and rely on CLI in the future.

If there is a chance `ensoup@2024.4.2` could execute `ensoup install engine` followed by `ensoup run` (which was broken last week), that'd be great.
